File: leetcode/double_pointer/11.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        vector<int>::iterator left, i, right, j;
        left = i = height.begin();
        right = j = height.end()-1;

        int max_area = 0;
        int current_area;
        while ( i != j ) {
            // i = max()
            if ( *left < *right ) {
                if ( *i > *left ) {
                    left = i;
                } else {
                    ++i;
                }
            } else {
                if ( *j > *right ) {
                    right = j;
                } else {
                    --j;
                }
            }

            current_area = min(*left, *right) * (right-left);
            max_area = max(max_area, current_area);
        }
        return max_area;
    }
};

File: leetcode/double_pointer/11.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
  // 1. 设置左右的最高
  // 2. 左右哪个比较低就优先移动哪个，因为水量是被左右的最低点限制的
  // 3. 持续监控水量，更新最大水量
  let [left, right] = [0, height.length-1];
  let volume = (right-left) * Math.min(height[left], height[right]);
  while ( left < right ) {
      if ( height[left] <= height[right] ) {
          ++left;
      } else {
          --right;
      }
      volume = Math.max(volume, (right-left) * Math.min(height[left], height[right]));
  }
  return volume;
};